src/cmds/pbs_rsub.c

```diff
@@ -696,8 +765,158 @@ main(int argc, char *argv[], char *envp[])
 		}
 	}
 
+	if (is_maintenance_resv) {
+		int i;
+
+		pbs_errno = 0;
+		bstat_head = pbs_stathost(connect, "", NULL, NULL);
+		if (bstat_head == NULL) {
+			if (pbs_errno) {
+				errmsg = pbs_geterrmsg(connect);
+				if (errmsg != NULL) {
+					fprintf(stderr, "pbs_rsub: %s\n", errmsg);
+				} else {
+					fprintf(stderr, "pbs_rsub: Error (%d) submitting reservation\n", pbs_errno);
+				}
+			} else {
+				fprintf(stderr, "pbs_rsub: No nodes found\n");
+			}
+
+			CS_close_app();
+			exit(pbs_errno);
+		}
+
+		/* here, the select is crafted from hosts */
+		hostp = maintenance_hosts;
+		for (;*hostp ;hostp++) {
+			/* lets craft select_str */
+			for (bstat = bstat_head; bstat; bstat = bstat->next) {
+				for (pattr = bstat->attribs; pattr; pattr = pattr->next) {
+				    if (pattr->resource && strcmp(pattr->name, "resources_available") == 0 && strcmp(pattr->resource, "ncpus") == 0)
+					ncpus = pattr->value;
+				}
+
+				if (strcmp(bstat->name, *hostp) == 0 && atoi(ncpus) > 0) {
+					if (!select_str) {
+						/* +20 stands for "select=host=" + ":ncpus=" + '\0' */
+						select_str = (char*) malloc(strlen(bstat->name) + strlen(ncpus) + 20);
+						if (select_str == NULL) {
+							fprintf(stderr, "pbs_rsub: Out of memory\n");
+							exit(2);
+						}
+						sprintf(select_str, "select=host=%s:ncpus=%s", bstat->name, ncpus);
+					} else {
+						tmp_str = strdup(select_str);
+						if (tmp_str == NULL) {
+							fprintf(stderr, "pbs_rsub: Out of memory\n");
+							exit(2);
+						}
+						/* +14 stands for "+host=" + ":ncpus=" + '\0' */
+						select_str = (char*) realloc(select_str, strlen(tmp_str) + strlen(bstat->name) + strlen(ncpus) + 14);
```

Comment:
I would suggest we initially malloc() a buffer of a fixed length say 500, and then use pbs_strcat() to concatenate subsequent values, instead of realloc()ing every time. This will save us from the calculations.

Reply:
Definitely, good suggestion, done.